### DISCUSSION TITLE
An example for a long-running publisher

### DISCUSSION THREAD

**Discussion by karls** [pika/pika#1382] [Ideas]
Hi,

I recently had to work through how to create a consumer-publisher pair for the platform I'm helping build. After setting it all up it seemed to work OK when there was lots of traffic on the upstream queue, but during periods of lower traffic the publisher would get its connection closed and the RabbitMQ (AWS MQ) logs would show a missed heartbeat failure.

I was super puzzled by that and spent several days going through the docs, examples, Googling, etc. It seems like there are [quite](https://stackoverflow.com/questions/73111224/connection-to-two-rabbitmq-servers/73194276) [a few](https://groups.google.com/g/pika-python/c/RFVHuF3GUCY) [people](https://groups.google.com/g/pika-python/c/M2IUG1ogpSI/m/tBjNJV_ZAwAJ) hitting the same or a similar issue.

What ended up being the issue was not realising that the publisher cannot just wait on an internal queue or some other mechanism for messages, as that would block the IO loop for the Pika connection, causing it to miss heartbeats. However, the [publish example](https://github.com/pika/pika/blob/main/examples/publish.py) doesn't in any way indicate how one would set up a long-running publisher — as part of a web-server/crawler/etc. I suspect many people take that example and modify it enough to get it to work for their purposes, often working around the closed connection issues with a `try...except` hack.

Do you think it'd be a good idea to include an example in `examples/` of how one would set up a long-running publisher? And maybe include relevant information in the docs for [`Connection.process_data_events`](https://pika.readthedocs.io/en/stable/modules/adapters/blocking.html#pika.adapters.blocking_connection.BlockingConnection.process_data_events)?

I've created [a repo](https://github.com/karls/rabbitmq-two-connections), which I included in my SO answer, which demonstrates — hopefully — the correct way to run a long-running publisher (feedback/tips very welcome). The code in the repo covers a bit more ground than is strictly needed for a self-contained example, but it could be adapted pretty easily.

**Comment by lukebakken** [comment]
@karls thanks for taking the time to set up an example. I'm assuming that when you state "long running publisher" you mean "slow publisher that blocks internally" which, yep, would block Pika's I/O loop.

What isn't clear in the docs with regard to [`process_data_events`?](https://pika.readthedocs.io/en/stable/modules/adapters/blocking.html#pika.adapters.blocking_connection.BlockingConnection.process_data_events). I agree, it's succinct, but it does state what the function does.

Feel free to submit a PR for adding your code as well as expanding the documentation to make it clearer how to use `BlockingConnection` in a scenario where your code blocks elsewhere.

**Comment by karls** [comment]
Hey @lukebakken.

> you mean "slow publisher that blocks internally"

Pretty much yep. Long-running, waiting for data on a queue/webhook/whatever. I suppose if it publishes very slowly there isn't really a need to keep the connection open. Just in my case the traffic was very variable.

> What isn't clear in the docs with regard to [process_data_events?](https://pika.readthedocs.io/en/stable/modules/adapters/blocking.html#pika.adapters.blocking_connection.BlockingConnection.process_data_events)

I guess if I had been more intimately familiar with how AMQP works I'd have spotted this earlier. From the docstring it's not clear why should I care about data events (what are they even?) being processed. Again, it's very likely down to my own ignorance, but I suspect there are many people like me who read some docs, see that a connection should be opened and kept open, develop a mental model of how it should work, and are then puzzled that it doesn't quite do that.
Given the importance of `process_data_events`, I think I would've expected to see an example that uses this method or the docstring being more explicit about having to call this method in certain situations.
On the other hand, a method docstring _should_ be succinct, so I don't know if an elaboration on why/how/etc is a good fit there. An example would work best and maybe linking from the docstring?

I'll put together a PR with an example and maybe a few tweaks for the docstring and we can collaborate in the PR from there.

**Comment by karls** [comment]
Hey @lukebakken 

Apologies for the delay, couldn't find the time to sit down and flesh this out.

I've created #1384 with an example and an updated docstring for `process_data_events`.

**Reply by lukebakken** [reply]
No rush! Thanks! I'll check it out when I have time.

**Comment by shouldsee** [comment]
Hi @karls,

This example is really useful and completely ~solved~ reduced my issues with publisher/consumer suffering from "missed heartbeats" and "ConnectionResetError(104, 'Connection reset by peer')"

```
 _AsyncTransportBase._initate_abort(): Initiating abrupt asynchronous transport shutdown: state=1; error=ConnectionResetError(104, 'Connection reset by peer'); <socket.socket fd=9, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('10.150.0.9', 40402)>
```

Thanks so much for sharing!

Kind regards
Feng

**Reply by karls** [reply]
Thank you @shouldsee! I'm glad you found it useful.
